### PR TITLE
Shutdown unix streams before dropping

### PR DESCRIPTION
**Pull request by A6GibKm**

### CODE REVIEW

**Comment by A6GibKm** [comment]
cc @SeaDve  this fixes the issue in Bustle 0.9.

**Comment by SeaDve** [comment]
> cc @SeaDve this fixes the issue in Bustle 0.9.

Interesting. Still, there's probably an issue somewhere, causing bustle/zbus to freeze but not bustle/dbus-monitor.

**Comment by A6GibKm** [comment]
I think shutdown will write the eof byte and we might be just waiting forever if it was not set. What i dont follow is that close should implicitly call shutdown.

File: src/desktop/secret.rs [hunk 1]
```diff
@@ -93,11 +93,23 @@ impl<'a> std::ops::Deref for Secret<'a> {
...
+        let (x1, mut x2) = UnixStream::pair()?;
+        proxy.retrieve(&x2).await?;
+        x2.shutdown().await?;
+        x1
+    };
+    #[cfg(feature = "async-std")]
+    let mut x1 = {
+        let (x1, x2) = UnixStream::pair()?;
+        proxy.retrieve(&x2).await?;
+        x2.shutdown(Shutdown::Both)?;
```
Comment (A6GibKm) [inline src/desktop/secret.rs:109]: Afaik, there is no need to close both halves, closing the writing half should suffice.

**Comment by A6GibKm** [comment]
@SeaDve @bilelmoussaoui Could you please verify that this actually fixes the issue?

The reproducer is to call the retrieve method while bustle 0.9 (note that flathub is at 0.10) is recording the session bus.

**Comment by bilelmoussaoui** [comment]
> @SeaDve @bilelmoussaoui Could you please verify that this actually fixes the issue?
> 
> The reproducer is to call the retrieve method while bustle 0.9 (note that flathub is at 0.10) is recording the session bus.

I have it on my todo list. will try to have a look the upcoming days

**Comment by A6GibKm** [comment]
Can confirm that this fixes the issues when tested with ASHPD demo (async-std) and Bustle on the commit before migrating to dbus-monitor and without this MR it does not work.

**Pull request merged**